src/garage/torch/algos/vpg.py

```diff
@@ -0,0 +1,230 @@
+            if not stop_entropy_gradient:
+                raise ValueError('stop_gradient should be True when '
+                                 'entropy_method is max')
+        if entropy_method == 'no_entropy':
+            if policy_ent_coeff != 0.0:
+                raise ValueError('policy_ent_coeff should be zero '
+                                 'when there is no entropy method')
+
+    def train_once(self, itr, paths):
+        """Perform one step of policy optimization."""
+        samples_data = self.process_samples(itr, paths)
+
+        self._optimize_policy(itr, paths)
+        self.baseline.fit(paths)
+
+        return samples_data['average_return']
+
+    def _optimize_policy(self, itr, paths):
+        self.optimizer.zero_grad()
+
+        policy_entropies = torch.stack([
+            self._add_padding(
+                self.policy.get_entropy(torch.Tensor(path['observations'])),
+                self.max_path_length)  # noqa: E501
+            for path in paths
+        ])
+
+        valids = [len(path['actions']) for path in paths]
+
+        baselines = torch.stack([
+            self._add_padding(self._get_baselines(path), self.max_path_length)
+            for path in paths
+        ])
+
+        rewards = torch.stack([
+            self._add_padding(torch.Tensor(path['rewards']),
+                              self.max_path_length)  # noqa: E501
+            for path in paths
+        ])
+
+        if self.maximum_entropy:
+            rewards += self.policy_ent_coeff * policy_entropies
+
+        deltas = (rewards + self.discount * F.pad(baselines,
+                                                  (0, 1))[:, 1:] - baselines)
+        deltas = F.pad(deltas,
+                       (0, self.max_path_length - 1)).unsqueeze(0).unsqueeze(0)
+
+        advantages = F.conv2d(deltas, self.filter, stride=1).squeeze()
+
+        if self.center_adv:
+            means, vars = list(
+                zip(*[(valid_adv.mean(), valid_adv.var())
+                      for valid_adv in self._filter_valids(advantages, valids)]
+                    ))  # noqa: E501
+            advantages = F.batch_norm(advantages.t(),
+                                      torch.Tensor(means),
+                                      torch.Tensor(vars),
+                                      eps=self.eps).t()
+
+        if self.positive_adv:
+            advantages -= advantages.min()
+
+        log_likelihoods = torch.stack([
+            self._add_padding(
+                self.policy.get_log_likelihood(
+                    torch.Tensor(path['observations']),
+                    torch.Tensor(path['actions'])), self.max_path_length)
+            for path in paths
+        ])
+
+        outputs = advantages * log_likelihoods
+
+        if self.entropy_regularzied:
+            outputs += self.policy_ent_coeff * policy_entropies
+
+        loss = torch.cat(self._filter_valids(outputs, valids))
+        (-loss).mean().backward()
```

Comment:
this seems like an important step. perhaps you should add a comment for it.

Reply:
add comments 
-----------
using a negative because optimizers use gradient descent, whilst we want gradient ascent.
-----------